File: Data Structures & Algorithms/combination-target-sum-ii/submission-19.js
class Solution {
    constructor() {
        this.res = [];
        this.count = new Map();
    }

    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(nums, target) {
        const cur = [];
        const A = [];
        
        for (const num of nums) {
            if (!this.count.has(num)) {
                A.push(num);
            }
            this.count.set(num, (this.count.get(num) || 0) + 1);
        }
        this.backtrack(A, target, cur, 0);
        return this.res;
    }

     /**
     * @param {number[]} nums
     * @param {number} target
     * @param {number[]} cur
     * @param {number} i
     * @return {void}
     */
    backtrack(nums, target, cur, i) {
        if (target === 0) {
            this.res.push([...cur]);
            return;
        }
        if (target < 0 || i >= nums.length) {
            return;
        }

        if (this.count.get(nums[i]) > 0) {
            cur.push(nums[i]);
            this.count.set(nums[i], this.count.get(nums[i]) - 1);
            this.backtrack(nums, target - nums[i], cur, i);
            this.count.set(nums[i], this.count.get(nums[i]) + 1);
            cur.pop();
        }

        this.backtrack(nums, target, cur, i + 1);
    }
}

File: Data Structures & Algorithms/target-sum/submission-19.cs
public class Solution {
    private int[,] dp;
    private int totalSum;

    public int FindTargetSumWays(int[] nums, int target) {
        totalSum = 0;
        foreach (var num in nums) totalSum += num;
        dp = new int[nums.Length, 2 * totalSum + 1];
        for (int i = 0; i < nums.Length; i++) {
            for (int j = 0; j < 2 * totalSum + 1; j++) {
                dp[i, j] = int.MinValue;  
            }
        }
        return Backtrack(0, 0, nums, target);
    }

    private int Backtrack(int i, int total, int[] nums, int target) {
        if (i == nums.Length) {
            return total == target ? 1 : 0;
        }

        if (dp[i, total + totalSum] != int.MinValue) {
            return dp[i, total + totalSum];  
        }

        dp[i, total + totalSum] = Backtrack(i + 1, total + nums[i], nums, target) + 
                                  Backtrack(i + 1, total - nums[i], nums, target);  
        return dp[i, total + totalSum];
    }
}